streams/src/main/java/org/apache/kafka/streams/kstream/Consumed.java

```diff
@@ -99,8 +99,8 @@ protected Consumed(final Consumed<K, V> consumed) {
     public static <K, V> Consumed<K, V> with(final Serde<K> keySerde,
                                              final Serde<V> valueSerde,
                                              final TimestampExtractor timestampExtractor,
-                                             final Topology.AutoOffsetReset resetPolicy) {
```

Comment:
We cannot just change this method, but need to keep it as-is, but only mark as `@Deprecated` and add a new overload with take the new `AutoOffsetReset` class instead of the old enum.

(Internally, we can translate the old enum to the new class right away...)

Same below.

(This comment does only apply to this class, for `ConsumedInternal` the change is ok, as it's an internal class and we can just change it right away.)

Reply:
Okay, so if I am understanding this before I make a change; we will have two 'versions', the original (which I am to deprecate), and the newer one I've implemented? Why was the way I did this incorrect?

Reply:
>  we will have two 'versions', the original (which I am to deprecate), and the newer one I've implemented? 

Correct.

> Why was the way I did this incorrect?

We need to preserve backward compatibility. If there is an existing Kafka Streams program using eg 3.9 release, users should be able to just change the dependency from 3.9 to 4.0, and recompile their program w/o the need to change any code. -- If we would apply the change as you did it originally however, their code would not compile any longer (if the use this method), and changes to their program would be necessary to compile using Kafka Streams 4.0, and thus it would be an breaking change, and we cannot have any breaking changes.

Does this make sense?

Reply:
Okay - this makes good sense to me. How/when does the old deprecated version get removed in this case then? Is there a point where a deprecated change gets cleaned out? 

I am working now to fix - redacting the change I made and refactoring to two 'versions', as you pointed out was correct. Thank you for helping me understand

Reply:
In general (ie, for most software project), deprecated APIs are only removed in major releases, so for this case, maybe in Apache Kafka 5.0.0. (We are currently removing a lot of previously deprecated APIs for the upcoming 4.0.0 release.)

For Kafka, we also keep deprecated APIs for at least 3 minor release, ie, if something is deprecated in Kafka 3.7/3.8/3.9, we cannot remove it in 4.0.0 release but keep it as well. This is a Kafka specific approach, and different projects do this differently.

Reply:
Okay, so one day the old deprecated methods will get cleared out, giving current users time to adjust what they might need to. I am new to the industry and my teams software is still pre-release, I have not come across this scenario yet. I appreciate you explaining it to me